drivers/flash/soc_flash_nrf.h

```diff
@@ -0,0 +1,160 @@
+ * @retval @ref FLASH_OP_DONE once operation was done, @ref FLASH_OP_ONGOING if
+ *         opperation needs more time for execution.
+ */
+typedef int (*flash_op_handler_t) (void *context);
+
+struct flash_op_desc {
+	flash_op_handler_t handler;
+	struct flash_context *context; /* [in,out] */
+};
+
+/**
+ * Synchronization backend driver initialization procedure.
+ *
+ * This will be run within flash driver initialization
+ */
+int nrfsb_synch_backend_init(void);
+
+/**
+ * Give synchronization context for write operation.
+ *
+ * @retval Synchronization context for the write operation.
+ * Structure of contex depends on backend implementation.
+ */
+void *nrfsb_get_write_synch_context(void);
+
+/**
+ * Give synchronization context for erase operation.
+ *
+ * @retval Synchronization context for the erase operation.
+ */
+void *nrfsb_get_erase_synch_context(void);
+
+/**
+ * Check if the operation need to be run synchonous with radio.
+ *
+ * @retval True if operation need to be run synchronously, otherwise False
+ */
+bool nrfsb_is_synchronization_required(void);
+
+/**
+ * Execute the flash opperation synchronously along the radio operations.
+ *
+ * Function executes calbacks op_desc->handler() in execution windows according
+ * to timing settings transfereed by op_desc->context->synch_context.
+ * This routine need to be called the handler as many time as it returns
+ * FLASH_OP_ONGOING, howewer an operation timeot should be implemented.
+ * When the handler() returns FLASH_OP_DONE, no further execution windows are
+ * needed so function should return as the handler() finished its operation.
+ *
+ * @retval 0 if op_desc->handler() was executed and
+ * finnished its opperation. Otherwise (timeout, couldn't schedule execution...)
+ * a negative errorcode.
+ */
+int nrfsb_work_synchronous_with_radio(struct flash_op_desc *op_desc);
+
+/**
+ * @}
+ */
+
+/**
+ * @defgroup nrf_flash_synch_timing synch timing backend API
+ * @ingroup nrf_flash_synch
+ * @{
+ *
+ * API which is used by nrf flash driver for check where execution fill in
+ * the execution window.
+ *
+ * API is used as fallow:
+ * begin of execution window
+ * get_timestamp_begin()
+ * [does some chunk of work]
+ * check_estimated_time_limit() == 0
+ * [does some chunk of work]
+ * check_estimated_time_limit() == 0
+ * [does some chunk of work]
+ * ...
+ * check_estimated_time_limit() == 0
+ * [preserve work context for next execution window]
+ * [return from execution window]
+ * end of execution window
+ */
+
+/**
+ * Get timestamp and store it in synchronization backend
+ * contex data as opperation beginning time reference.
+ * This timesapm will be used by @ref check_estimated_time_limit()
+ * as the execution window begin referece.
+ *
+ * @param *synch_context timing contex to store the timestamp.
+ *                       Structure of contex depends on backend implementation.
+ */
+void nrfsb_get_timestamp_begin(void *synch_context);
```

Comment:
It seems that this function and `nrfsb_check_estimated_time_limit` could be optional capability of synchronization context. If these functions were not implemented, each chunk of flash operation would be called for consecutive `flash_op_handler_t handler()` calls, what would have worse performance, but implementation would be easier.

We could consider making it optional to implement synchronization backends in steps:
1. PoC without this capability.
2. Final optimal solution with this capability.

I think that for some scheduling algorithms used by synchronization backends, this capability makes no sense and could be omitted even in final solution.